<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Welcome to Biopython_en’s documentation! &mdash; Biopython_en 1.0 documentation</title>
    
    <link rel="stylesheet" href="../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="top" title="Biopython_en 1.0 documentation" href="../../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../../index.html">Biopython_en 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="welcome-to-biopython-en-s-documentation">
<h1>Welcome to Biopython_en&#8217;s documentation!<a class="headerlink" href="#welcome-to-biopython-en-s-documentation" title="Permalink to this headline">¶</a></h1>
<p>Contents:</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="chr1.html">Chapter 1  Introduction</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr1.html#what-is-biopython">1.1  What is Biopython?</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr1.html#what-can-i-find-in-the-biopython-package">1.2  What can I find in the Biopython package</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr1.html#installing-biopython">1.3  Installing Biopython</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr1.html#frequently-asked-questions-faq">1.4  Frequently Asked Questions (FAQ)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr2.html">Chapter 2  Quick Start – What can you do with Biopython?</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr2.html#general-overview-of-what-biopython-provides">2.1  General overview of what Biopython provides</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr2.html#working-with-sequences">2.2  Working with sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr2.html#a-usage-example">2.3  A usage example</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr2.html#parsing-sequence-file-formats">2.4  Parsing sequence file formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr2.html#connecting-with-biological-databases">2.5  Connecting with biological databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr2.html#what-to-do-next">2.6  What to do next</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr3.html">Chapter 3  Sequence objects</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#sequences-and-alphabets">3.1  Sequences and Alphabets</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#sequences-act-like-strings">3.2  Sequences act like strings</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#slicing-a-sequence">3.3  Slicing a sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#turning-seq-objects-into-strings">3.4  Turning Seq objects into strings</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#concatenating-or-adding-sequences">3.5  Concatenating or adding sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#changing-case">3.6  Changing case</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#nucleotide-sequences-and-reverse-complements">3.7  Nucleotide sequences and (reverse) complements</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#transcription">3.8  Transcription</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#translation">3.9  Translation</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#translation-tables">3.10  Translation Tables</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#comparing-seq-objects">3.11  Comparing Seq objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#mutableseq-objects">3.12  MutableSeq objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#unknownseq-objects">3.13  UnknownSeq objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr3.html#working-with-directly-strings">3.14  Working with directly strings</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr4.html">Chapter 4  Sequence annotation objects</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr4.html#the-seqrecord-object">4.1  The SeqRecord object</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr4.html#creating-a-seqrecord">4.2  Creating a SeqRecord</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr4.html#feature-location-and-position-objects">4.3  Feature, location and position objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr4.html#references">4.4  References</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr4.html#the-format-method">4.5  The format method</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr4.html#slicing-a-seqrecord">4.6  Slicing a SeqRecord</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr4.html#adding-seqrecord-objects">4.7  Adding SeqRecord objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr4.html#reverse-complementing-seqrecord-objects">4.8  Reverse-complementing SeqRecord objects</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr5.html">Chapter 5  Sequence Input/Output</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr5.html#parsing-or-reading-sequences">5.1  Parsing or Reading Sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr5.html#parsing-sequences-from-compressed-files">5.2  Parsing sequences from compressed files</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr5.html#parsing-sequences-from-the-net">5.3  Parsing sequences from the net</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr5.html#sequence-files-as-dictionaries">5.4  Sequence files as Dictionaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr5.html#writing-sequence-files">5.5  Writing Sequence Files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr6.html">Chapter 6  Multiple Sequence Alignment objects</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr6.html#parsing-or-reading-sequence-alignments">6.1  Parsing or Reading Sequence Alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr6.html#writing-alignments">6.2  Writing Alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr6.html#manipulating-alignments">6.3  Manipulating Alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr6.html#alignment-tools">6.4  Alignment Tools</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr7.html">Chapter 7  BLAST</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr7.html#running-blast-over-the-internet">7.1  Running BLAST over the Internet</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr7.html#running-blast-locally">7.2  Running BLAST locally</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr7.html#parsing-blast-output">7.3  Parsing BLAST output</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr7.html#the-blast-record-class">7.4  The BLAST record class</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr7.html#deprecated-blast-parsers">7.5  Deprecated BLAST parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr7.html#dealing-with-psi-blast">7.6  Dealing with PSI-BLAST</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr7.html#dealing-with-rps-blast">7.7  Dealing with RPS-BLAST</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr8.html">Chapter 8  BLAST and other sequence search tools (<em>experimental code</em>)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr8.html#the-searchio-object-model">8.1  The SearchIO object model</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr8.html#a-note-about-standards-and-conventions">8.2  A note about standards and conventions</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr8.html#reading-search-output-files">8.3  Reading search output files</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr8.html#dealing-with-large-search-output-files-with-indexing">8.4  Dealing with large search output files with indexing</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr8.html#writing-and-converting-search-output-files">8.5  Writing and converting search output files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr9.html">hapter 9  Accessing NCBI’s Entrez databases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#entrez-guidelines">9.1  Entrez Guidelines</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#einfo-obtaining-information-about-the-entrez-databases">9.2  EInfo: Obtaining information about the Entrez databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#esearch-searching-the-entrez-databases">9.3  ESearch: Searching the Entrez databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#epost-uploading-a-list-of-identifiers">9.4  EPost: Uploading a list of identifiers</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#esummary-retrieving-summaries-from-primary-ids">9.5  ESummary: Retrieving summaries from primary IDs</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#efetch-downloading-full-records-from-entrez">9.6  EFetch: Downloading full records from Entrez</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#elink-searching-for-related-items-in-ncbi-entrez">9.7  ELink: Searching for related items in NCBI Entrez</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#egquery-global-query-counts-for-search-terms">9.8  EGQuery: Global Query - counts for search terms</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#espell-obtaining-spelling-suggestions">9.9  ESpell: Obtaining spelling suggestions</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#parsing-huge-entrez-xml-files">9.10  Parsing huge Entrez XML files</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#handling-errors">9.11  Handling errors</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#specialized-parsers">9.12  Specialized parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#using-a-proxy">9.13  Using a proxy</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#examples">9.14  Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr9.html#using-the-history-and-webenv">9.15  Using the history and WebEnv</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr10.html">Chapter 10  Swiss-Prot and ExPASy</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr10.html#parsing-swiss-prot-files">10.1  Parsing Swiss-Prot files</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr10.html#parsing-prosite-records">10.2  Parsing Prosite records</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr10.html#parsing-prosite-documentation-records">10.3  Parsing Prosite documentation records</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr10.html#parsing-enzyme-records">10.4  Parsing Enzyme records</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr10.html#accessing-the-expasy-server">10.5  Accessing the ExPASy server</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr10.html#scanning-the-prosite-database">10.6  Scanning the Prosite database</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr11.html">Chapter 11  Going 3D: The PDB module</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr11.html#reading-and-writing-crystal-structure-files">11.1  Reading and writing crystal structure files</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr11.html#structure-representation">11.2  Structure representation</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr11.html#disorder">11.3  Disorder</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr11.html#hetero-residues">11.4  Hetero residues</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr11.html#navigating-through-a-structure-object">11.5  Navigating through a Structure object</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr11.html#analyzing-structures">11.6  Analyzing structures</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr11.html#common-problems-in-pdb-files">11.7  Common problems in PDB files</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr11.html#accessing-the-protein-data-bank">11.8  Accessing the Protein Data Bank</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr11.html#general-questions">11.9  General questions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr12.html">Chapter 12  Bio.PopGen: Population genetics</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr12.html#genepop">12.1  GenePop</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr12.html#coalescent-simulation">12.2  Coalescent simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr12.html#other-applications">12.3  Other applications</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr12.html#future-developments">12.4  Future Developments</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr13.html">Chapter 13  Phylogenetics with Bio.Phylo</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr13.html#demo-whats-in-a-tree">13.1  Demo: What’s in a Tree?</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr13.html#i-o-functions">13.2  I/O functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr13.html#view-and-export-trees">13.3  View and export trees</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr13.html#using-tree-and-clade-objects">13.4  Using Tree and Clade objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr13.html#running-external-applications">13.5  Running external applications</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr13.html#paml-integration">13.6  PAML integration</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr13.html#future-plans">13.7  Future plans</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr14.html">Chapter 14  Sequence motif analysis using Bio.motifs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr14.html#motif-objects">14.1  Motif objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr14.html#position-weight-matrices">14.2  Position-Weight Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr14.html#position-specific-scoring-matrices">14.3  Position-Specific Scoring Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr14.html#searching-for-instances">14.4  Searching for instances</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr14.html#each-motif-object-has-an-associated-position-specific-scoring-matrix">14.5  Each motif object has an associated Position-Specific Scoring Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr14.html#comparing-motifs">14.6  Comparing motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr14.html#de-novo-motif-finding">14.7  <em>De novo</em> motif finding</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr14.html#useful-links">14.8  Useful links</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr14.html#obsolete-bio-motif-module">14.9  Obsolete Bio.Motif module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr15.html">Chapter 15  Cluster analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#data-representation">Data representation</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#missing-values">Missing values</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#random-number-generator">Random number generator</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#euclidean-distance">Euclidean distance</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#city-block-distance">City-block distance</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#the-pearson-correlation-coefficient">The Pearson correlation coefficient</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#absolute-pearson-correlation">Absolute Pearson correlation</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#uncentered-correlation-cosine-of-the-angle">Uncentered correlation (cosine of the angle)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#absolute-uncentered-correlation">Absolute uncentered correlation</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#spearman-rank-correlation">Spearman rank correlation</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#kendalls">Kendall’s τ</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#weighting">Weighting</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#calculating-the-distance-matrix">Calculating the distance matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#calculating-the-cluster-centroids">Calculating the cluster centroids</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#calculating-the-distance-between-clusters">Calculating the distance between clusters</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#k-means-and-k-medians"><em>k</em>-means and <em>k</em>-medians</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#k-medoids-clustering"><em>k</em>-medoids clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#representing-a-hierarchical-clustering-solution">Representing a hierarchical clustering solution</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#performing-hierarchical-clustering">Performing hierarchical clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#id1">Calculating the distance matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#id2">Calculating the cluster centroids</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#id3">Calculating the distance between clusters</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#id4">Performing hierarchical clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#performing-k-means-or-k-medians-clustering">Performing <em>k</em>-means or <em>k</em>-medians clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#calculating-a-self-organizing-map">Calculating a Self-Organizing Map</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr15.html#saving-the-clustering-result">Saving the clustering result</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr16.html">Chapter 16  Supervised learning methods</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr16.html#the-logistic-regression-model">16.1  The Logistic Regression Model</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr16.html#k-nearest-neighbors">16.2  <em>k</em>-Nearest Neighbors</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr16.html#naive-bayes">16.3  Naïve Bayes</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr16.html#maximum-entropy">16.4  Maximum Entropy</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr16.html#markov-models">16.5  Markov Models</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr17.html">Chapter 17  Graphics including GenomeDiagram</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr17.html#genomediagram">17.1  GenomeDiagram</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr17.html#chromosomes">17.2  Chromosomes</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr18.html">Chapter 18  Cookbook – Cool things to do with it</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr18.html#working-with-sequence-files">18.1  Working with sequence files</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr18.html#sequence-parsing-plus-simple-plots">18.2  Sequence parsing plus simple plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr18.html#dealing-with-alignments">18.3  Dealing with alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr18.html#substitution-matrices">18.4  Substitution Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr18.html#biosql-storing-sequences-in-a-relational-database">18.5  BioSQL – storing sequences in a relational database</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr19.html">Chapter 19  The Biopython testing framework</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr19.html#running-the-tests">19.1  Running the tests</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr19.html#writing-tests">19.2  Writing tests</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr19.html#writing-doctests">19.3  Writing doctests</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chr20.html">Chapter 20  Advanced</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chr20.html#parser-design">20.1  Parser Design</a></li>
<li class="toctree-l2"><a class="reference internal" href="chr20.html#substitution-matrices">20.2  Substitution Matrices</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="../../genindex.html"><em>Index</em></a></li>
<li><a class="reference internal" href="../../py-modindex.html"><em>Module Index</em></a></li>
<li><a class="reference internal" href="../../search.html"><em>Search Page</em></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Welcome to Biopython_en&#8217;s documentation!</a><ul>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../../_sources/en_test/source/index.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../../index.html">Biopython_en 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Biopython.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
  </body>
</html>