<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Chapter 14 Sequence motif analysis using Bio.motifs &mdash; Biopython_en 1.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="Biopython_en 1.0 documentation" href="../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">Biopython_en 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="chapter-14-sequence-motif-analysis-using-bio-motifs">
<h1>Chapter 14  Sequence motif analysis using Bio.motifs<a class="headerlink" href="#chapter-14-sequence-motif-analysis-using-bio-motifs" title="Permalink to this headline">¶</a></h1>
<p>This chapter gives an overview of the functionality of the
<tt class="docutils literal"><span class="pre">Bio.motifs</span></tt> package included in Biopython. It is intended for people
who are involved in the analysis of sequence motifs, so I’ll assume that
you are familiar with basic notions of motif analysis. In case something
is unclear, please look at Section <a class="reference external" href="#sec:links">14.8</a> for some
relevant links.</p>
<p>Most of this chapter describes the new <tt class="docutils literal"><span class="pre">Bio.motifs</span></tt> package included
in Biopython 1.61 onwards, which is replacing the older <tt class="docutils literal"><span class="pre">Bio.Motif</span></tt>
package introduced with Biopython 1.50, which was in turn based on two
older former Biopython modules, <tt class="docutils literal"><span class="pre">Bio.AlignAce</span></tt> and <tt class="docutils literal"><span class="pre">Bio.MEME</span></tt>. It
provides most of their functionality with a unified motif object
implementation.</p>
<p>Speaking of other libraries, if you are reading this you might be
interested in <a class="reference external" href="http://fraenkel.mit.edu/TAMO/">TAMO</a>, another python
library designed to deal with sequence motifs. It supports more
<em>de-novo</em> motif finders, but it is not a part of Biopython and has some
restrictions on commercial use.</p>
<div class="section" id="motif-objects">
<h2>14.1  Motif objects<a class="headerlink" href="#motif-objects" title="Permalink to this headline">¶</a></h2>
<p>Since we are interested in motif analysis, we need to take a look at
<tt class="docutils literal"><span class="pre">Motif</span></tt> objects in the first place. For that we need to import the
Bio.motifs library:</p>
<p>and we can start creating our first motif objects. We can either create
a <tt class="docutils literal"><span class="pre">Motif</span></tt> object from a list of instances of the motif, or we can
obtain a <tt class="docutils literal"><span class="pre">Motif</span></tt> object by parsing a file from a motif database or
motif finding software.</p>
<div class="section" id="creating-a-motif-from-instances">
<h3>14.1.1  Creating a motif from instances<a class="headerlink" href="#creating-a-motif-from-instances" title="Permalink to this headline">¶</a></h3>
<p>Suppose we have these instances of a DNA motif:</p>
<p>then we can create a Motif object as follows:</p>
<p>The instances are saved in an attribute <tt class="docutils literal"><span class="pre">m.instances</span></tt>, which is
essentially a Python list with some added functionality, as described
below. Printing out the Motif object shows the instances from which it
was constructed:</p>
<p>The length of the motif defined as the sequence length, which should be
the same for all instances:</p>
<p>The Motif object has an attribute <tt class="docutils literal"><span class="pre">.counts</span></tt> containing the counts of
each nucleotide at each position. Printing this counts matrix shows it
in an easily readable format:</p>
<p>You can access these counts as a dictionary:</p>
<p>but you can also think of it as a 2D array with the nucleotide as the
first dimension and the position as the second dimension:</p>
<p>You can also directly access columns of the counts matrix</p>
<p>Instead of the nucleotide itself, you can also use the index of the
nucleotide in the sorted letters in the alphabet of the motif:</p>
<p>The motif has an associated consensus sequence, defined as the sequence
of letters along the positions of the motif for which the largest value
in the corresponding columns of the <tt class="docutils literal"><span class="pre">.counts</span></tt> matrix is obtained:</p>
<p>as well as an anticonsensus sequence, corresponding to the smallest
values in the columns of the <tt class="docutils literal"><span class="pre">.counts</span></tt> matrix:</p>
<p>You can also ask for a degenerate consensus sequence, in which ambiguous
nucleotides are used for positions where there are multiple nucleotides
with high counts:</p>
<p>Here, W and R follow the IUPAC nucleotide ambiguity codes: W is either A
or T, and V is A, C, or G [<a class="reference external" href="#cornish1985">10</a>]. The degenerate
consensus sequence is constructed following the rules specified by
Cavener [<a class="reference external" href="#cavener1987">11</a>].</p>
<p>We can also get the reverse complement of a motif:</p>
<p>The reverse complement and the degenerate consensus sequence are only
defined for DNA motifs.</p>
</div>
<div class="section" id="reading-motifs">
<h3>14.1.2  Reading motifs<a class="headerlink" href="#reading-motifs" title="Permalink to this headline">¶</a></h3>
<p>Creating motifs from instances by hand is a bit boring, so it’s useful
to have some I/O functions for reading and writing motifs. There are no
really well established standards for storing motifs, but there’s a
couple of formats which are more used than others. The most important
distinction is whether the motif representation is based on instances or
on some version of PWM matrix.</p>
<div class="section" id="jaspar">
<h4>JASPAR<a class="headerlink" href="#jaspar" title="Permalink to this headline">¶</a></h4>
<p>One of the most popular motif databases
<a class="reference external" href="http://jaspar.genereg.net">JASPAR</a> stores motifs either as a list of
instances, or as a frequency matrix. As an example, these are the
beginning and ending lines of the JASPAR <tt class="docutils literal"><span class="pre">Arnt.sites</span></tt> file showing
known binding sites of the mouse helix-loop-helix transcription factor
Arnt:</p>
<p>The parts of the sequence in capital letters are the motif instances
that were found to align to each other.</p>
<p>We can create a <tt class="docutils literal"><span class="pre">Motif</span></tt> object from these instances as follows:</p>
<p>The instances from which this motif was created is stored in the
<tt class="docutils literal"><span class="pre">.instances</span></tt> property:</p>
<p>The counts matrix of this motif is automatically calculated from the
instances:</p>
<p>The JASPAR database also makes motifs available directly as a count
matrix, without the instances from which it was created. For example,
this is the JASPAR file <tt class="docutils literal"><span class="pre">SRF.pfm</span></tt> containing the count matrix for the
human SRF transcription factor:</p>
<p>We can create a motif for this count matrix as follows:</p>
<p>As this motif was created from the counts matrix directly, it has no
instances associated with it:</p>
<p>We can now ask for the consensus sequence of these two motifs:</p>
</div>
<div class="section" id="meme">
<h4>MEME<a class="headerlink" href="#meme" title="Permalink to this headline">¶</a></h4>
<p>MEME [<a class="reference external" href="#bailey1994">12</a>] is a tool for discovering motifs in a
group of related DNA or protein sequences. It takes as input a group of
DNA or protein sequences and outputs as many motifs as requested.
Therefore, in contrast to JASPAR files, MEME output files typically
contain multiple motifs. This is an example.</p>
<p>At the top of an output file generated by MEME shows some background
information about the MEME and the version of MEME used:</p>
<p>Further down, the input set of training sequences is recapitulated:</p>
<p>and the exact command line that was used:</p>
<p>Next is detailed information on each motif that was found:</p>
<p>To parse this file (stored as <tt class="docutils literal"><span class="pre">meme.dna.oops.txt</span></tt>), use</p>
<p>The <tt class="docutils literal"><span class="pre">motifs.parse</span></tt> command reads the complete file directly, so you
can close the file after calling <tt class="docutils literal"><span class="pre">motifs.parse</span></tt>. The header
information is stored in attributes:</p>
<p>The record is an object of the <tt class="docutils literal"><span class="pre">Bio.motifs.meme.Record</span></tt> class. The
class inherits from list, and you can think of <tt class="docutils literal"><span class="pre">record</span></tt> as a list of
Motif objects:</p>
<p>In addition to these generic motif attributes, each motif also stores
its specific information as calculated by MEME. For example,</p>
<p>In addition to using an index into the record, as we did above, you can
also find it by its name:</p>
<p>Each motif has an attribute <tt class="docutils literal"><span class="pre">.instances</span></tt> with the sequence instances
in which the motif was found, providing some information on each
instance:</p>
</div>
<div class="section" id="mast">
<h4>MAST<a class="headerlink" href="#mast" title="Permalink to this headline">¶</a></h4>
</div>
<div class="section" id="transfac">
<h4>TRANSFAC<a class="headerlink" href="#transfac" title="Permalink to this headline">¶</a></h4>
<p>TRANSFAC is a manually curated database of transcription factors,
together with their genomic binding sites and DNA binding profiles
[<a class="reference external" href="#matys2003">27</a>]. While the file format used in the TRANSFAC
database is nowadays also used by others, we will refer to it as the
TRANSFAC file format.</p>
<p>A minimal file in the TRANSFAC format looks as follows:</p>
<p>This file shows the frequency matrix of motif <tt class="docutils literal"><span class="pre">motif1</span></tt> of 12
nucleotides. In general, one file in the TRANSFAC format can contain
multiple motifs. For example, this is the contents of the example
TRANSFAC file <tt class="docutils literal"><span class="pre">transfac.dat</span></tt>:</p>
<p>To parse a TRANSFAC file, use</p>
<p>The overall version number, if available, is stored as
<tt class="docutils literal"><span class="pre">record.version</span></tt>:</p>
<p>Each motif in <tt class="docutils literal"><span class="pre">record</span></tt> is in instance of the
<tt class="docutils literal"><span class="pre">Bio.motifs.transfac.Motif</span></tt> class, which inherits both from the
<tt class="docutils literal"><span class="pre">Bio.motifs.Motif</span></tt> class and from a Python dictionary. The dictionary
uses the two-letter keys to store any additional information about the
motif:</p>
<p>TRANSFAC files are typically much more elaborate than this example,
containing lots of additional information about the motif. Table
<a class="reference external" href="#table:transfaccodes">14.1.2</a> lists the two-letter field codes that
are commonly found in TRANSFAC files:</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>Table 14.1: Fields commonly found in TRANSFAC files</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="84%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">AC</span></tt></td>
<td>Accession number</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">AS</span></tt></td>
<td>Accession numbers, secondary</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">BA</span></tt></td>
<td>Statistical basis</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">BF</span></tt></td>
<td>Binding factors</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">BS</span></tt></td>
<td>Factor binding sites underlying the matrix</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">CC</span></tt></td>
<td>Comments</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">CO</span></tt></td>
<td>Copyright notice</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">DE</span></tt></td>
<td>Short factor description</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">DR</span></tt></td>
<td>External databases</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">DT</span></tt></td>
<td>Date created/updated</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">HC</span></tt></td>
<td>Subfamilies</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">HP</span></tt></td>
<td>Superfamilies</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">ID</span></tt></td>
<td>Identifier</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">NA</span></tt></td>
<td>Name of the binding factor</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">OC</span></tt></td>
<td>Taxonomic classification</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">OS</span></tt></td>
<td>Species/Taxon</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">OV</span></tt></td>
<td>Older version</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">PV</span></tt></td>
<td>Preferred version</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">TY</span></tt></td>
<td>Type</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">XX</span></tt></td>
<td>Empty line; these are not stored in the Record.</td>
</tr>
</tbody>
</table>
</div></blockquote>
<p>Each motif also has an attribute <tt class="docutils literal"><span class="pre">.references</span></tt> containing the
references associated with the motif, using these two-letter keys:</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>Table 14.2: Fields used to store references in TRANSFAC files</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="32%" />
<col width="68%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">RN</span></tt></td>
<td>Reference number</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">RA</span></tt></td>
<td>Reference authors</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">RL</span></tt></td>
<td>Reference data</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">RT</span></tt></td>
<td>Reference title</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">RX</span></tt></td>
<td>PubMed ID</td>
</tr>
</tbody>
</table>
</div></blockquote>
<p>Printing the motifs writes them out in their native TRANSFAC format:</p>
<p>You can export the motifs in the TRANSFAC format by capturing this
output in a string and saving it in a file:</p>
</div>
</div>
<div class="section" id="writing-motifs">
<h3>14.1.3  Writing motifs<a class="headerlink" href="#writing-motifs" title="Permalink to this headline">¶</a></h3>
<p>Speaking of exporting, let’s look at export functions in general. To
export a motif in the JASPAR <tt class="docutils literal"><span class="pre">.pfm</span></tt> format, use</p>
<p>To write the motif in a TRANSFAC-like matrix format, use</p>
<p>To write out multiple motifs, you can use <tt class="docutils literal"><span class="pre">motifs.write</span></tt>. This
function can be used regardless of whether the motifs originated from a
TRANSFAC file. For example,</p>
</div>
<div class="section" id="creating-a-sequence-logo">
<h3>14.1.4  Creating a sequence logo<a class="headerlink" href="#creating-a-sequence-logo" title="Permalink to this headline">¶</a></h3>
<p>If we have internet access, we can create a
<a class="reference external" href="http://weblogo.berkeley.edu">weblogo</a>:</p>
<p>We should get our logo saved as a PNG in the specified file.</p>
</div>
</div>
<div class="section" id="position-weight-matrices">
<h2>14.2  Position-Weight Matrices<a class="headerlink" href="#position-weight-matrices" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">.counts</span></tt> attribute of a Motif object shows how often each
nucleotide appeared at each position along the alignment. We can
normalize this matrix by dividing by the number of instances in the
alignment, resulting in the probability of each nucleotide at each
position along the alignment. We refer to these probabilities as the
position-weight matrix. However, beware that in the literature this term
may also be used to refer to the position-specific scoring matrix, which
we discuss below.</p>
<p>Usually, pseudocounts are added to each position before normalizing.
This avoids overfitting of the position-weight matrix to the limited
number of motif instances in the alignment, and can also prevent
probabilities from becoming zero. To add a fixed pseudocount to all
nucleotides at all positions, specify a number for the <tt class="docutils literal"><span class="pre">pseudocounts</span></tt>
argument:</p>
<p>Alternatively, <tt class="docutils literal"><span class="pre">pseudocounts</span></tt> can be a dictionary specifying the
pseudocounts for each nucleotide. For example, as the GC content of the
human genome is about 40%, you may want to choose the pseudocounts
accordingly:</p>
<p>The position-weight matrix has its own methods to calculate the
consensus, anticonsensus, and degenerate consensus sequences:</p>
<p>Note that due to the pseudocounts, the degenerate consensus sequence
calculated from the position-weight matrix is slightly different from
the degenerate consensus sequence calculated from the instances in the
motif:</p>
<p>The reverse complement of the position-weight matrix can be calculated
directly from the <tt class="docutils literal"><span class="pre">pwm</span></tt>:</p>
</div>
<div class="section" id="position-specific-scoring-matrices">
<h2>14.3  Position-Specific Scoring Matrices<a class="headerlink" href="#position-specific-scoring-matrices" title="Permalink to this headline">¶</a></h2>
<p>Using the background distribution and PWM with pseudo-counts added, it’s
easy to compute the log-odds ratios, telling us what are the log odds of
a particular symbol to be coming from a motif against the background. We
can use the <tt class="docutils literal"><span class="pre">.log_odds()</span></tt> method on the position-weight matrix:</p>
<p>Here we can see positive values for symbols more frequent in the motif
than in the background and negative for symbols more frequent in the
background. 0.0 means that it’s equally likely to see a symbol in the
background and in the motif.</p>
<p>This assumes that A, C, G, and T are equally likely in the background.
To calculate the position-specific scoring matrix against a background
with unequal probabilities for A, C, G, T, use the <tt class="docutils literal"><span class="pre">background</span></tt>
argument. For example, against a background with a 40% GC content, use</p>
<p>The maximum and minimum score obtainable from the PSSM are stored in the
<tt class="docutils literal"><span class="pre">.max</span></tt> and <tt class="docutils literal"><span class="pre">.min</span></tt> properties:</p>
<p>The mean and standard deviation of the PSSM scores with respect to a
specific background are calculated by the <tt class="docutils literal"><span class="pre">.mean</span></tt> and <tt class="docutils literal"><span class="pre">.std</span></tt>
methods.</p>
<p>A uniform background is used if <tt class="docutils literal"><span class="pre">background</span></tt> is not specified. The
mean is particularly important, as its value is equal to the
Kullback-Leibler divergence or relative entropy, and is a measure for
the information content of the motif compared to the background. As in
Biopython the base-2 logarithm is used in the calculation of the
log-odds scores, the information content has units of bits.</p>
<p>The <tt class="docutils literal"><span class="pre">.reverse_complement</span></tt>, <tt class="docutils literal"><span class="pre">.consensus</span></tt>, <tt class="docutils literal"><span class="pre">.anticonsensus</span></tt>, and
<tt class="docutils literal"><span class="pre">.degenerate_consensus</span></tt> methods can be applied directly to PSSM
objects.</p>
</div>
<div class="section" id="searching-for-instances">
<h2>14.4  Searching for instances<a class="headerlink" href="#searching-for-instances" title="Permalink to this headline">¶</a></h2>
<p>The most frequent use for a motif is to find its instances in some
sequence. For the sake of this section, we will use an artificial
sequence like this:</p>
<div class="section" id="searching-for-exact-matches">
<h3>14.4.1  Searching for exact matches<a class="headerlink" href="#searching-for-exact-matches" title="Permalink to this headline">¶</a></h3>
<p>The simplest way to find instances, is to look for exact matches of the
true instances of the motif:</p>
<p>We can do the same with the reverse complement (to find instances on the
complementary strand):</p>
</div>
<div class="section" id="searching-for-matches-using-the-pssm-score">
<h3>14.4.2  Searching for matches using the PSSM score<a class="headerlink" href="#searching-for-matches-using-the-pssm-score" title="Permalink to this headline">¶</a></h3>
<p>It’s just as easy to look for positions, giving rise to high log-odds
scores against our motif:</p>
<p>The negative positions refer to instances of the motif found on the
reverse strand of the test sequence, and follow the Python convention on
negative indices. Therefore, the instance of the motif at <tt class="docutils literal"><span class="pre">pos</span></tt> is
located at <tt class="docutils literal"><span class="pre">test_seq[pos:pos+len(m)]</span></tt> both for positive and for
negative values of <tt class="docutils literal"><span class="pre">pos</span></tt>.</p>
<p>You may notice the threshold parameter, here set arbitrarily to 3.0.
This is in <em>log</em><sub>2</sub>, so we are now looking only for words, which
are eight times more likely to occur under the motif model than in the
background. The default threshold is 0.0, which selects everything that
looks more like the motif than the background.</p>
<p>You can also calculate the scores at all positions along the sequence:</p>
<p>In general, this is the fastest way to calculate PSSM scores. The scores
returned by <tt class="docutils literal"><span class="pre">pssm.calculate</span></tt> are for the forward strand only. To
obtain the scores on the reverse strand, you can take the reverse
complement of the PSSM:</p>
</div>
<div class="section" id="selecting-a-score-threshold">
<h3>14.4.3  Selecting a score threshold<a class="headerlink" href="#selecting-a-score-threshold" title="Permalink to this headline">¶</a></h3>
<p>If you want to use a less arbitrary way of selecting thresholds, you can
explore the distribution of PSSM scores. Since the space for a score
distribution grows exponentially with motif length, we are using an
approximation with a given precision to keep computation cost
manageable:</p>
<p>The <tt class="docutils literal"><span class="pre">distribution</span></tt> object can be used to determine a number of
different thresholds. We can specify the requested false-positive rate
(probability of “finding” a motif instance in background generated
sequence):</p>
<p>or the false-negative rate (probability of “not finding” an instance
generated from the motif):</p>
<p>or a threshold (approximately) satisfying some relation between the
false-positive rate and the false-negative rate (fnr/fpr≃ <em>t</em>):</p>
<p>or a threshold satisfying (roughly) the equality between the
false-positive rate and the −<em>log</em> of the information content (as used
in patser software by Hertz and Stormo):</p>
<p>For example, in case of our motif, you can get the threshold giving you
exactly the same results (for this sequence) as searching for instances
with balanced threshold with rate of 1000.</p>
</div>
</div>
<div class="section" id="each-motif-object-has-an-associated-position-specific-scoring-matrix">
<h2>14.5  Each motif object has an associated Position-Specific Scoring Matrix<a class="headerlink" href="#each-motif-object-has-an-associated-position-specific-scoring-matrix" title="Permalink to this headline">¶</a></h2>
<p>To facilitate searching for potential TFBSs using PSSMs, both the
position-weight matrix and the position-specific scoring matrix are
associated with each motif. Using the Arnt motif as an example:</p>
<p>The negative infinities appear here because the corresponding entry in
the frequency matrix is 0, and we are using zero pseudocounts by
default:</p>
<p>If you change the <tt class="docutils literal"><span class="pre">.pseudocounts</span></tt> attribute, the position-frequency
matrix and the position-specific scoring matrix are recalculated
automatically:</p>
<p>You can also set the <tt class="docutils literal"><span class="pre">.pseudocounts</span></tt> to a dictionary over the four
nucleotides if you want to use different pseudocounts for them. Setting
<tt class="docutils literal"><span class="pre">motif.pseudocounts</span></tt> to <tt class="docutils literal"><span class="pre">None</span></tt> resets it to its default value of
zero.</p>
<p>The position-specific scoring matrix depends on the background
distribution, which is uniform by default:</p>
<p>Again, if you modify the background distribution, the position-specific
scoring matrix is recalculated:</p>
<p>Setting <tt class="docutils literal"><span class="pre">motif.background</span></tt> to <tt class="docutils literal"><span class="pre">None</span></tt> resets it to a uniform
distribution:</p>
<p>If you set <tt class="docutils literal"><span class="pre">motif.background</span></tt> equal to a single value, it will be
interpreted as the GC content:</p>
<p>Note that you can now calculate the mean of the PSSM scores over the
background against which it was computed:</p>
<p>as well as its standard deviation:</p>
<p>and its distribution:</p>
<p>Note that the position-weight matrix and the position-specific scoring
matrix are recalculated each time you call <tt class="docutils literal"><span class="pre">motif.pwm</span></tt> or
<tt class="docutils literal"><span class="pre">motif.pssm</span></tt>, respectively. If speed is an issue and you want to use
the PWM or PSSM repeatedly, you can save them as a variable, as in</p>
</div>
<div class="section" id="comparing-motifs">
<h2>14.6  Comparing motifs<a class="headerlink" href="#comparing-motifs" title="Permalink to this headline">¶</a></h2>
<p>Once we have more than one motif, we might want to compare them.</p>
<p>Before we start comparing motifs, I should point out that motif
boundaries are usually quite arbitrary. This means we often need to
compare motifs of different lengths, so comparison needs to involve some
kind of alignment. This means we have to take into account two things:</p>
<ul class="simple">
<li>alignment of motifs</li>
<li>some function to compare aligned motifs</li>
</ul>
<p>To align the motifs, we use ungapped alignment of PSSMs and substitute
zeros for any missing columns at the beginning and end of the matrices.
This means that effectively we are using the background distribution for
columns missing from the PSSM. The distance function then returns the
minimal distance between motifs, as well as the corresponding offset in
their alignment.</p>
<p>To give an exmaple, let us first load another motif, which is similar to
our test motif <tt class="docutils literal"><span class="pre">m</span></tt>:</p>
<p>To make the motifs comparable, we choose the same values for the
pseudocounts and the background distribution as our motif <tt class="docutils literal"><span class="pre">m</span></tt>:</p>
<p>We’ll compare these motifs using the Pearson correlation. Since we want
it to resemble a distance measure, we actually take 1−<em>r</em>, where <em>r</em>
is the Pearson correlation coefficient (PCC):</p>
<p>This means that the best PCC between motif <tt class="docutils literal"><span class="pre">m</span></tt> and <tt class="docutils literal"><span class="pre">m_reb1</span></tt> is
obtained with the following alignment:</p>
<p>where <tt class="docutils literal"><span class="pre">b</span></tt> stands for background distribution. The PCC itself is
roughly 1−0.239=0.761.</p>
</div>
<div class="section" id="de-novo-motif-finding">
<h2>14.7  <em>De novo</em> motif finding<a class="headerlink" href="#de-novo-motif-finding" title="Permalink to this headline">¶</a></h2>
<p>Currently, Biopython has only limited support for <em>de novo</em> motif
finding. Namely, we support running and parsing of AlignAce and MEME.
Since the number of motif finding tools is growing rapidly,
contributions of new parsers are welcome.</p>
<div class="section" id="id1">
<h3>14.7.1  MEME<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>Let’s assume, you have run MEME on sequences of your choice with your
favorite parameters and saved the output in the file <tt class="docutils literal"><span class="pre">meme.out</span></tt>. You
can retrieve the motifs reported by MEME by running the following piece
of code:</p>
<p>Besides the most wanted list of motifs, the result object contains more
useful information, accessible through properties with self-explanatory
names:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">.alphabet</span></tt></li>
<li><tt class="docutils literal"><span class="pre">.datafile</span></tt></li>
<li><tt class="docutils literal"><span class="pre">.sequence_names</span></tt></li>
<li><tt class="docutils literal"><span class="pre">.version</span></tt></li>
<li><tt class="docutils literal"><span class="pre">.command</span></tt></li>
</ul>
<p>The motifs returned by the MEME Parser can be treated exactly like
regular Motif objects (with instances), they also provide some extra
functionality, by adding additional information about the instances.</p>
</div>
<div class="section" id="alignace">
<h3>14.7.2  AlignAce<a class="headerlink" href="#alignace" title="Permalink to this headline">¶</a></h3>
<p>We can do very similar things with the AlignACE program. Assume, you
have your output in the file <tt class="docutils literal"><span class="pre">alignace.out</span></tt>. You can parse your output
with the following code:</p>
<p>Again, your motifs behave as they should:</p>
<p>In fact you can even see, that AlignAce found a very similar motif as
MEME. It is just a longer version of a reverse complement of the MEME
motif:</p>
<p>If you have AlignAce installed on the same machine, you can also run it
directly from Biopython. A short example of how this can be done is
shown below (other parameters can be specified as keyword parameters):</p>
<p>Since AlignAce prints all of its output to standard output, you can get
to your motifs by parsing the first part of the result:</p>
</div>
</div>
<div class="section" id="useful-links">
<h2>14.8  Useful links<a class="headerlink" href="#useful-links" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><a class="reference external" href="http://en.wikipedia.org/wiki/Sequence_motif">Sequence motif</a> in
wikipedia</li>
<li><a class="reference external" href="http://en.wikipedia.org/wiki/Position_weight_matrix">PWM</a> in
wikipedia</li>
<li><a class="reference external" href="http://en.wikipedia.org/wiki/Consensus_sequence">Consensus
sequence</a> in
wikipedia</li>
<li><a class="reference external" href="http://bio.cs.washington.edu/assessment/">Comparison of different motif finding
programs</a></li>
</ul>
</div>
<div class="section" id="obsolete-bio-motif-module">
<h2>14.9  Obsolete Bio.Motif module<a class="headerlink" href="#obsolete-bio-motif-module" title="Permalink to this headline">¶</a></h2>
<p>The rest of this chapter above describes the <tt class="docutils literal"><span class="pre">Bio.motifs</span></tt> package
included in Biopython 1.61 onwards, which is replacing the older
<tt class="docutils literal"><span class="pre">Bio.Motif</span></tt> package introduced with Biopython 1.50, which was in turn
based on two older former Biopython modules, <tt class="docutils literal"><span class="pre">Bio.AlignAce</span></tt> and
<tt class="docutils literal"><span class="pre">Bio.MEME</span></tt>.</p>
<p>To allow for a smooth transition, the older <tt class="docutils literal"><span class="pre">Bio.Motif</span></tt> package will
be maintained in parallel with its replacement <tt class="docutils literal"><span class="pre">Bio.motifs</span></tt> at least
two more releases, and at least one year.</p>
<div class="section" id="id2">
<h3>14.9.1  Motif objects<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>Since we are interested in motif analysis, we need to take a look at
<tt class="docutils literal"><span class="pre">Motif</span></tt> objects in the first place. For that we need to import the
Motif library:</p>
<p>and we can start creating our first motif objects. Let’s create a DNA
motif:</p>
<p>This is for now just an empty container, so let’s add some sequences to
our newly created motif:</p>
<p>Now we have a full <tt class="docutils literal"><span class="pre">Motif</span></tt> instance, so we can try to get some basic
information about it. Let’s start with length and consensus sequence:</p>
<p>In case of DNA motifs, we can also get a reverse complement of a motif:</p>
<p>We can also calculate the information content of a motif with a simple
call:</p>
<p>This gives us a number of bits of information provided by the motif,
which tells us how much differs from background.</p>
<p>The most common representation of a motif is a PWM (Position Weight
Matrix). It summarizes the probabilities of finding any symbol (in this
case nucleotide) in any position of a motif. It can be computed by
calling the <tt class="docutils literal"><span class="pre">.pwm()</span></tt> method:</p>
<p>The probabilities in the motif’s PWM are based on the counts in the
instances, but we can see, that even though there were no Gs and no Cs
in the instances, we still have non-zero probabilities assigned to them.
These come from pseudo-counts which are, roughly speaking, a commonly
used way to acknowledge the incompleteness of our knowledge and avoid
technical problems with calculating logarithms of 0.</p>
<p>We can control the way that pseudo-counts are added with two properties
of Motif objects <tt class="docutils literal"><span class="pre">.background</span></tt> is the probability distribution over
all symbols in the alphabet that we assume represents background,
non-motif sequences (usually based on the GC content of the respective
genome). It is by default set to a uniform distribution upon creation of
a motif:</p>
<p>The other parameter is <tt class="docutils literal"><span class="pre">.beta</span></tt>, which states the amount of
pseudo-counts we should add to the PWM. By default it is set to 1.0,</p>
<p>so that the total input of pseudo-counts is equal to that of one
instance.</p>
<p>Using the background distribution and pwm with pseudo-counts added, it’s
easy to compute the log-odds ratios, telling us what are the log odds of
a particular symbol to be coming from a motif against the background. We
can use the <tt class="docutils literal"><span class="pre">.log_odds()</span></tt> method:</p>
<p>Here we can see positive values for symbols more frequent in the motif
than in the background and negative for symbols more frequent in the
background. 0.0 means that it’s equally likely to see a symbol in
background and in the motif (e.g. ‘T’ in the second-last position).</p>
<div class="section" id="reading-and-writing">
<h4>14.9.1.1  Reading and writing<a class="headerlink" href="#reading-and-writing" title="Permalink to this headline">¶</a></h4>
<p>Creating motifs from instances by hand is a bit boring, so it’s useful
to have some I/O functions for reading and writing motifs. There are no
really well established standards for storing motifs, but there’s a
couple of formats which are more used than others. The most important
distinction is whether the motif representation is based on instances or
on some version of PWM matrix. On of the most popular motif databases
<a class="reference external" href="http://jaspar.genereg.net">JASPAR</a> stores motifs in both formats, so
let’s look at how we can import JASPAR motifs from instances:</p>
<p>and from a count matrix:</p>
<p>The <tt class="docutils literal"><span class="pre">arnt</span></tt> and <tt class="docutils literal"><span class="pre">srf</span></tt> motifs can both do the same things for us, but
they use different internal representations of the motif. We can tell
that by inspecting the <tt class="docutils literal"><span class="pre">has_counts</span></tt> and <tt class="docutils literal"><span class="pre">has_instances</span></tt> properties:</p>
<p>There are conversion functions, which can help us convert between
different representations:</p>
<p>The important thing to remember here is that the method
<tt class="docutils literal"><span class="pre">make_instances_from_counts()</span></tt> creates fake instances, because usually
there are very many possible sets of instances which give rise to the
same pwm, and if we have only the count matrix, we cannot reconstruct
the original one. This does not make any difference if we are using the
PWM as the representation of the motif, but one should be careful with
exporting instances from count-based motifs.</p>
<p>Speaking of exporting, let’s look at export functions. We can export to
fasta:</p>
<p>or to TRANSFAC-like matrix format (used by some motif processing
software)</p>
<p>Finally, if we have internet access, we can create a
<a class="reference external" href="http://weblogo.berkeley.edu">weblogo</a>:</p>
<p>We should get our logo saved as a png in the specified file.</p>
</div>
</div>
<div class="section" id="id3">
<h3>14.9.2  Searching for instances<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<p>The most frequent use for a motif is to find its instances in some
sequence. For the sake of this section, we will use an artificial
sequence like this:</p>
<p>The simplest way to find instances, is to look for exact matches of the
true instances of the motif:</p>
<p>We can do the same with the reverse complement (to find instances on the
complementary strand):</p>
<p>It’s just as easy to look for positions, giving rise to high log-odds
scores against our motif:</p>
<p>You may notice the threshold parameter, here set arbitrarily to 5.0.
This is in <em>log</em><sub>2</sub>, so we are now looking only for words, which
are 32 times more likely to occur under the motif model than in the
background. The default threshold is 0.0, which selects everything that
looks more like the motif than the background.</p>
<p>If you want to use a less arbitrary way of selecting thresholds, you can
explore the <tt class="docutils literal"><span class="pre">Motif.score_distribution</span></tt> class implementing an
distribution of scores for a given motif. Since the space for a score
distribution grows exponentially with motif length, we are using an
approximation with a given precision to keep computation cost
manageable:</p>
<p>The sd object can be used to determine a number of different thresholds.</p>
<p>We can specify the requested false-positive rate (probability of
“finding” a motif instance in background generated sequence):</p>
<p>or the false-negative rate (probability of “not finding” an instance
generated from the motif):</p>
<p>or a threshold (approximately) satisfying some relation between fpr and
fnr <em>fnr</em>/<em>fpr</em>≃ <em>t</em>:</p>
<p>or a threshold satisfying (roughly) the equality between the
false-positive rate and the −<em>log</em> of the information content (as used
in patser software by Hertz and Stormo).</p>
<p>For example, in case of our motif, you can get the threshold giving you
exactly the same results (for this sequence) as searching for instances
with balanced threshold with rate of 1000.</p>
</div>
<div class="section" id="id4">
<h3>14.9.3  Comparing motifs<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<p>Once we have more than one motif, we might want to compare them. For
that, we have currently three different methods of <tt class="docutils literal"><span class="pre">Bio.Motif</span></tt>
objects.</p>
<p>Before we start comparing motifs, I should point out that motif
boundaries are usually quite arbitrary. This means, that we often need
to compare motifs of different lengths, so comparison needs to involve
some kind of alignment. This means, that we have to take into account
two things:</p>
<ul class="simple">
<li>alignment of motifs</li>
<li>some function to compare aligned motifs</li>
</ul>
<p>In <tt class="docutils literal"><span class="pre">Bio.Motif</span></tt> we have 3 different functions for motif comparison,
which are based on the same idea behind motif alignment, but use
different functions to compare aligned motifs. Briefly speaking, we are
using ungapped alignment of PWMs and substitute the missing columns at
the beginning and end of the matrices with background distribution. All
three comparison functions are written in such a way, that they can be
interpreted as distance measures, however only one (<tt class="docutils literal"><span class="pre">dist_dpq</span></tt>)
satisfies the triangle inequality. All of them return the minimal
distance and the corresponding offset between motifs.</p>
<p>To show how these functions work, let us first load another motif, which
is similar to our test motif <tt class="docutils literal"><span class="pre">m</span></tt>:</p>
<p>The first function we’ll use to compare these motifs is based on Pearson
correlation. Since we want it to resemble a distance measure, we
actually take 1−<em>r</em>, where <em>r</em> is the Pearson correlation coefficient
(PCC):</p>
<p>This means, that the best PCC between motif <tt class="docutils literal"><span class="pre">m</span></tt> and <tt class="docutils literal"><span class="pre">Ubx</span></tt> is
obtained with the following alignment:</p>
<p>where <tt class="docutils literal"><span class="pre">b</span></tt> stands for background distribution. The PCC itself is
roughly 1−0.42=0.58. If we try the reverse complement of the Ubx motif:</p>
<p>We can see that the PCC is better (almost 0.75), and the alignment is
also different:</p>
<p>There are two other functions: <tt class="docutils literal"><span class="pre">dist_dpq</span></tt>, which is a true metric
(satisfying traingle inequality) based on the Kullback-Leibler
divergence</p>
<p>and the <tt class="docutils literal"><span class="pre">dist_product</span></tt> method, which is based on the product of
probabilities which can be interpreted as the probability of
independently generating the same instance by both motifs.</p>
</div>
<div class="section" id="id5">
<h3>14.9.4  <em>De novo</em> motif finding<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<p>Currently, Biopython has only limited support for <em>de novo</em> motif
finding. Namely, we support running and parsing of AlignAce and MEME.
Since the number of motif finding tools is growing rapidly,
contributions of new parsers are welcome.</p>
<div class="section" id="id6">
<h4>14.9.4.1  MEME<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h4>
<p>Let’s assume, you have run MEME on sequences of your choice with your
favorite parameters and saved the output in the file <tt class="docutils literal"><span class="pre">meme.out</span></tt>. You
can retrieve the motifs reported by MEME by running the following piece
of code:</p>
<p>Besides the most wanted list of motifs, the result object contains more
useful information, accessible through properties with self-explanatory
names:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">.alphabet</span></tt></li>
<li><tt class="docutils literal"><span class="pre">.datafile</span></tt></li>
<li><tt class="docutils literal"><span class="pre">.sequence_names</span></tt></li>
<li><tt class="docutils literal"><span class="pre">.version</span></tt></li>
<li><tt class="docutils literal"><span class="pre">.command</span></tt></li>
</ul>
<p>The motifs returned by MEMEParser can be treated exactly like regular
Motif objects (with instances), they also provide some extra
functionality, by adding additional information about the instances.</p>
</div>
<div class="section" id="id7">
<h4>14.9.4.2  AlignAce<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h4>
<p>We can do very similar things with AlignACE program. Assume, you have
your output in the file <tt class="docutils literal"><span class="pre">alignace.out</span></tt>. You can parse your output with
the following code:</p>
<p>Again, your motifs behave as they should:</p>
<p>In fact you can even see, that AlignAce found a very similar motif as
MEME, it is just a longer version of a reverse complement of MEME motif:</p>
<p>If you have AlignAce installed on the same machine, you can also run it
directly from Biopython. Short example of how this can be done is shown
below (other parameters can be specified as keyword parameters):</p>
<p>Since AlignAce prints all its output to standard output, you can get to
your motifs by parsing the first part of the result:</p>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Chapter 14  Sequence motif analysis using Bio.motifs</a><ul>
<li><a class="reference internal" href="#motif-objects">14.1  Motif objects</a><ul>
<li><a class="reference internal" href="#creating-a-motif-from-instances">14.1.1  Creating a motif from instances</a></li>
<li><a class="reference internal" href="#reading-motifs">14.1.2  Reading motifs</a><ul>
<li><a class="reference internal" href="#jaspar">JASPAR</a></li>
<li><a class="reference internal" href="#meme">MEME</a></li>
<li><a class="reference internal" href="#mast">MAST</a></li>
<li><a class="reference internal" href="#transfac">TRANSFAC</a></li>
</ul>
</li>
<li><a class="reference internal" href="#writing-motifs">14.1.3  Writing motifs</a></li>
<li><a class="reference internal" href="#creating-a-sequence-logo">14.1.4  Creating a sequence logo</a></li>
</ul>
</li>
<li><a class="reference internal" href="#position-weight-matrices">14.2  Position-Weight Matrices</a></li>
<li><a class="reference internal" href="#position-specific-scoring-matrices">14.3  Position-Specific Scoring Matrices</a></li>
<li><a class="reference internal" href="#searching-for-instances">14.4  Searching for instances</a><ul>
<li><a class="reference internal" href="#searching-for-exact-matches">14.4.1  Searching for exact matches</a></li>
<li><a class="reference internal" href="#searching-for-matches-using-the-pssm-score">14.4.2  Searching for matches using the PSSM score</a></li>
<li><a class="reference internal" href="#selecting-a-score-threshold">14.4.3  Selecting a score threshold</a></li>
</ul>
</li>
<li><a class="reference internal" href="#each-motif-object-has-an-associated-position-specific-scoring-matrix">14.5  Each motif object has an associated Position-Specific Scoring Matrix</a></li>
<li><a class="reference internal" href="#comparing-motifs">14.6  Comparing motifs</a></li>
<li><a class="reference internal" href="#de-novo-motif-finding">14.7  <em>De novo</em> motif finding</a><ul>
<li><a class="reference internal" href="#id1">14.7.1  MEME</a></li>
<li><a class="reference internal" href="#alignace">14.7.2  AlignAce</a></li>
</ul>
</li>
<li><a class="reference internal" href="#useful-links">14.8  Useful links</a></li>
<li><a class="reference internal" href="#obsolete-bio-motif-module">14.9  Obsolete Bio.Motif module</a><ul>
<li><a class="reference internal" href="#id2">14.9.1  Motif objects</a><ul>
<li><a class="reference internal" href="#reading-and-writing">14.9.1.1  Reading and writing</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id3">14.9.2  Searching for instances</a></li>
<li><a class="reference internal" href="#id4">14.9.3  Comparing motifs</a></li>
<li><a class="reference internal" href="#id5">14.9.4  <em>De novo</em> motif finding</a><ul>
<li><a class="reference internal" href="#id6">14.9.4.1  MEME</a></li>
<li><a class="reference internal" href="#id7">14.9.4.2  AlignAce</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/en/chr14.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">Biopython_en 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Biopython.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
  </body>
</html>